### PR TITLE
Add React and the env-size-display component

### PR DESCRIPTION
**Pull request by hatched**
React has been added to the codebase as well as the first tested react component env-size-display.

### CODE REVIEW

**Comment by hatched** [comment]
#### To QA

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/2856/
Test FAILed.

File: .eslintrc [hunk 1]
```diff
@@ -1,6 +1,10 @@
 env:
   browser: true
   node: true
+ecmaFeatures:
+  jsx: true
+  defaultParams: true
+  objectLiteralShorthandProperties: false
```
Comment (huwshimi) [inline .eslintrc]: Yay!

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/2857/
Test PASSed.

File: jujugui/static/gui/src/app/components/env-size-display/test-env-size-display.js [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+
+function queryComponentSelector(component, selector, all) {
+  var queryFn = (all) ? 'querySelectorAll' : 'querySelector';
+  return component.getDOMNode()[queryFn](selector);
+}
+
+describe('EnvSizeDisplay', function() {
+
+  beforeAll(function(done) {
+    console.log(chai.config)
```
Comment (huwshimi) [inline jujugui/static/gui/src/app/components/env-size-display/test-env-size-display.js]: I'm assuming this was left by mistake.

File: jujugui/static/gui/src/app/components/env-size-display/test-env-size-display.js [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+    chai.config.showDiff = true;
+    chai.config.includeStack = true;
+    chai.config.truncateThreashold = 0;
```
Comment (huwshimi) [inline jujugui/static/gui/src/app/components/env-size-display/test-env-size-display.js]: Do we need all these configs if they didn't work to fix the diff issues?

Comment (huwshimi) [inline jujugui/static/gui/src/app/components/env-size-display/test-env-size-display.js]: Also, if we do need them, can we move them to some global config?

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/2859/
Test PASSed.

File: jujugui/static/gui/src/app/subapps/browser/browser.js [hunk 4]
```diff
@@ -407,7 +407,9 @@ YUI.add('subapp-browser', function(Y) {
     */
     _machine: function(metadata) {
       this._renderMachineViewPanelView(this.get('db'), this.get('env'));
-      this.get('environmentHeader').setSelectedTab('machines');
```
Comment (huwshimi) [inline jujugui/static/gui/src/app/subapps/browser/browser.js]: We don't need this when the flag is turned off?

Comment (huwshimi) [inline jujugui/static/gui/src/app/subapps/browser/browser.js]: OK, so this ONLY sets when the flag is OFF.

**Comment by huwshimi** [comment]
Looks good @hatched! QA OK :+1:

File: jujugui/static/gui/src/app/components/env-size-display/test-env-size-display.js (hunk 2 above)
Comment (hatched) [inline jujugui/static/gui/src/app/components/env-size-display/test-env-size-display.js]: Yeah oops, removed

File: jujugui/static/gui/src/app/app.js [hunk 5]
```diff
@@ -688,6 +690,39 @@ YUI.add('juju-gui', function(Y) {
...
+      changeState system.
+
+      @method changeState
+      @param {Object} state The state to change the view to.
+    */
+    changeState: function(state) {
+      this.fire('changeState', state);
+    },
+
+    _renderEnvSizeDisplay: function(serviceCount=0, machineCount=0) {
```
Comment (makyo) [inline jujugui/static/gui/src/app/app.js]: Docstring

File: jujugui/static/gui/src/app/app.js [hunk 6]
```diff
@@ -688,6 +690,39 @@ YUI.add('juju-gui', function(Y) {
...
+
+      @method parseURLState
+    */
+    parseURLState: function(req, res, next) {
+      this.state.loadRequest(req, '', {dispatch: false});
+      next();
+    },
+
+    /**
+      Passed to the components so that they can interact with the existing
```
Comment (makyo) [inline jujugui/static/gui/src/app/app.js]: What is passed?  This doesn't seem to match the function contents.

File: jujugui/static/gui/src/app/components/env-size-display/test-env-size-display.js (hunk 3 above)
Comment (hatched) [inline jujugui/static/gui/src/app/components/env-size-display/test-env-size-display.js]: Will remove them for now..will have to investigate this fix later.

File: jujugui/static/gui/src/app/subapps/browser/browser.js (hunk 4 above)
Comment (hatched) [inline jujugui/static/gui/src/app/subapps/browser/browser.js]: correct

File: jujugui/static/gui/src/app/assets/javascripts/bind-function-pollyfill.js [hunk 7]
```diff
@@ -0,0 +1,47 @@
+// Taken from https://github.com/facebook/react/issues/945#issuecomment-37329894
+
+// Pollyfill for function.prototype.bind
```
Comment (makyo) [inline jujugui/static/gui/src/app/assets/javascripts/bind-function-pollyfill.js]: trivial: polyfill

File: jujugui/static/gui/src/app/app.js (hunk 5 above)
Comment (hatched) [inline jujugui/static/gui/src/app/app.js]: done

File: jujugui/static/gui/src/app/app.js (hunk 6 above)
Comment (hatched) [inline jujugui/static/gui/src/app/app.js]: fixed

File: jujugui/static/gui/src/app/assets/javascripts/bind-function-pollyfill.js (hunk 7 above)
Comment (hatched) [inline jujugui/static/gui/src/app/assets/javascripts/bind-function-pollyfill.js]: fixed

**Comment by makyo** [comment]
:+1:

**Comment by hatched** [comment]
Thanks for the reviews! :shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/juju-gui-merge

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/2867/
Test PASSed.

**Pull request merged**